File: canal-youtube/kit-produccion/scripts/montar_video.js
// Monta el video final a partir de la hoja de montaje, las imágenes de escena,
// el audio de la narración y el SRT. Un plano por fila de la hoja.
//
//   node montar_video.js config.json
//
// La hoja de montaje ya está cuadrada al audio real (la genera timing_sheet.js
// a partir de la duración medida del MP3), así que los cortes caen donde caen
// las frases. Aquí no se recalcula nada: se obedece la hoja.
//
// Requiere ffmpeg. Si no está en el sistema:  npm install ffmpeg-static
const fs = require('fs');
const path = require('path');
const { execFileSync } = require('child_process');

const cfgPath = process.argv[2];
if (!cfgPath) { console.error('uso: node montar_video.js config.json'); process.exit(1); }
const cfg = JSON.parse(fs.readFileSync(cfgPath, 'utf8'));
const dir = path.dirname(path.resolve(cfgPath));
const R = p => path.resolve(dir, p);

const FPS   = cfg.fps || 25;
const [W,H] = (cfg.resolucion || '1920x1080').split('x').map(Number);
const CRF   = cfg.crf || 20;

// ---- ffmpeg -----------------------------------------------------------------
function buscarFfmpeg() {
  if (cfg.ffmpeg) return R(cfg.ffmpeg);
  try { return execFileSync('which', ['ffmpeg']).toString().trim(); } catch (_) {}
  for (const base of [dir, process.cwd(), __dirname]) {
    try { return require(require.resolve('ffmpeg-static', { paths: [base] })); } catch (_) {}
  }
  throw new Error('No encuentro ffmpeg. Instala con: npm install ffmpeg-static');
}
const FFMPEG = buscarFfmpeg();

// ---- hoja de montaje --------------------------------------------------------
// CSV con ';' y tiempos m:ss. La duración de la hoja manda sobre todo lo demás.
const seg = t => { const [m,s] = String(t).trim().split(':').map(Number); return m*60 + s; };
const filas = fs.readFileSync(R(cfg.hoja), 'utf8')
  .replace(/^﻿/, '').trim().split(/\r?\n/).slice(1)
  .map(l => l.split(';'))
  .filter(c => c.length >= 5)
  .map(c => ({ n: Number(c[0]), bloque: c[1], entra: seg(c[2]), sale: seg(c[3]), escena: c[5] || '' }));

// ---- imágenes ---------------------------------------------------------------
// Se espera un archivo por plano nombrado con su número (01.jpg, 02.png…).
// `rellenar` cicla las que haya para poder previsualizar el montaje antes de
// tener las 78 generadas — sirve para ver el ritmo, no para publicar.
const EXT = ['.jpg', '.jpeg', '.png', '.webp'];
const imgDir = R(cfg.imagenes);
const disponibles = fs.readdirSync(imgDir)
  .filter(f => EXT.includes(path.extname(f).toLowerCase()))
  .sort();
if (!disponibles.length) throw new Error(`No hay imágenes en ${imgDir}`);

const porPlano = new Map();
for (const f of disponibles) {
  const m = path.basename(f, path.extname(f)).match(/^(\d+)/);
  if (m) porPlano.set(Number(m[1]), f);
}

let faltan = 0;
function imagenDe(plano, i) {
  const exacta = porPlano.get(plano.n);
  if (exacta) return path.join(imgDir, exacta);
  faltan++;
  if (!cfg.rellenar) throw new Error(`Falta la imagen del plano ${plano.n}. Usa "rellenar": true para previsualizar.`);
  return path.join(imgDir, disponibles[i % disponibles.length]);
}

// ---- selección de planos ----------------------------------------------------
let planos = filas;
if (cfg.hasta) planos = planos.filter(p => p.entra < cfg.hasta);
if (!planos.length) throw new Error('La selección no deja ningún plano.');
const ultimo = planos[planos.length - 1];
const finVideo = cfg.hasta ? Math.min(ultimo.sale, cfg.hasta) : ultimo.sale;

// ---- filtros ----------------------------------------------------------------
// zoompan trabaja en pasos de píxel entero: si no se sobreescala antes, el
// movimiento va a saltos. Se escala a 4K, se hace el Ken Burns y se baja a 1080.
// d=1 da un fotograma de salida por cada uno de entrada, así `on` avanza a lo
// largo de todo el plano y el zoom es un rampeo continuo.
const ZOOM_MAX = cfg.zoom || 1.12;
function kenBurns(idx, dur) {
  const frames = Math.max(1, Math.round(dur * FPS));
  const paso = (ZOOM_MAX - 1) / frames;
  const dentro = idx % 2 === 0;
  const z = dentro
    ? `min(1+${paso.toFixed(8)}*on,${ZOOM_MAX})`
    : `max(${ZOOM_MAX}-${paso.toFixed(8)}*on,1)`;
  // Deriva lateral suave, alternando el lado para que no todos los planos
  // se muevan igual.
  const lado = idx % 4 < 2 ? 1 : -1;
  const x = `iw/2-(iw/zoom/2)+${lado}*(iw*0.04)*(on/${frames})`;
  const y = `ih/2-(ih/zoom/2)`;
  return `scale=${W*2}:-2,zoompan=z='${z}':d=1:x='${x}':y='${y}':s=${W}x${H}:fps=${FPS},setsar=1,format=yuv420p`;
}

const entradas = [];
const filtros = [];
planos.forEach((p, i) => {
  const dur = (cfg.hasta ? Math.min(p.sale, cfg.hasta) : p.sale) - p.entra;
  if (dur <= 0) return;
  entradas.push('-loop', '1', '-t', dur.toFixed(3), '-i', imagenDe(p, i));
  const k = entradas.filter(a => a === '-i').length - 1;
  filtros.push(`[${k}:v]${kenBurns(i, dur)}[v${k}]`);
});
const nV = filtros.length;
let cadena = filtros.join(';') + ';' +
  Array.from({ length: nV }, (_, k) => `[v${k}]`).join('') +
  `concat=n=${nV}:v=1:a=0[vcat]`;

// Los subtítulos se queman al final, sobre el montaje ya concatenado, para que
// no los deforme el zoom.
let salidaV = '[vcat]';
if (cfg.subtitulos !== false && cfg.srt) {
  const srt = path.relative(dir, R(cfg.srt)).replace(/\\/g, '/');
  const estilo = cfg.estiloSubs ||
    `FontName=DejaVu Sans,Fontsize=22,Bold=1,PrimaryColour=&H00FFFFFF,` +
    `OutlineColour=&H00000000,BorderStyle=1,Outline=2,Shadow=1,` +
    `Alignment=2,MarginV=60`;
  cadena += `;[vcat]subtitles=${srt}:force_style='${estilo}'[vsub]`;
  salidaV = '[vsub]';
}

// ---- audio ------------------------------------------------------------------
const idxAudio = entradas.filter(a => a === '-i').length;
const args = [
  '-y', ...entradas,
  '-i', R(cfg.audio),
  '-filter_complex', cadena,
  '-map', salidaV,
  '-map', `${idxAudio}:a`,
  '-t', finVideo.toFixed(3),
  '-c:v', 'libx264', '-preset', cfg.preset || 'medium', '-crf', String(CRF),
  '-pix_fmt', 'yuv420p', '-r', String(FPS),
  '-c:a', 'aac', '-b:a', '192k',
  '-movflags', '+faststart',
  R(cfg.salida),
];

console.log(`ffmpeg: ${FFMPEG}`);
console.log(`planos: ${nV}${faltan ? `  (${faltan} sin imagen propia, rellenados)` : ''}`);
console.log(`duración: ${finVideo.toFixed(1)}s  ·  ${W}x${H} @ ${FPS}fps`);
execFileSync(FFMPEG, args, { stdio: ['ignore', 'ignore', 'inherit'], cwd: dir });

const mb = fs.statSync(R(cfg.salida)).size / 1e6;
console.log(`\n${cfg.salida} — ${mb.toFixed(1)} MB`);

File: canal-youtube/producciones/08-kaczynski-harvard/montar-video.py
"""Monta el video entero: 95 imagenes + narracion, con los tiempos de la hoja.

Uso:
    python3 montar-video.py CARPETA_IMAGENES [--zoom] [--salida video.mp4]

CARPETA_IMAGENES debe contener 001.png ... 095.png (tambien vale .jpg).

Lee hoja-montaje.csv, que ya trae la duracion exacta de cada plano anclada a
la narracion, y se lo pasa a ffmpeg por el demuxer concat. El audio se mezcla
tal cual, sin recodificar la voz mas de lo imprescindible.

--zoom anade un zoom lento tipo Ken Burns a cada imagen. Es lo que da
sensacion de movimiento con material fijo. Cuesta bastante mas de render
(cada imagen pasa por zoompan), asi que conviene montar primero sin el para
revisar el sincronismo y activarlo en la pasada final.

--subtitulos quema subtitulos.srt en la imagen. Se quema y no se adjunta
aparte porque el 80% del consumo del canal es en movil, donde los subtitulos
opcionales vienen apagados por defecto.

El estilo va escrito aqui y no en el SRT: cuerpo grande, blanco con borde
negro grueso y sombra, pegado a la parte baja pero no al borde, que es lo que
aguanta sobre cualquier imagen sin tener que mirar plano a plano.

CUIDADO CON EL ORDEN DE fps Y subtitles
Con el demuxer concat cada imagen entra como UN SOLO fotograma que dura todo
el plano. Si se quema antes de fijar la cadencia, el filtro dibuja sobre ese
fotograma unico el subtitulo activo en ese instante y luego -r duplica el
fotograma YA COMPUESTO durante todo el plano: cada plano acaba enseñando un
subtitulo congelado, el que estaba activo cuando entro la imagen. Se vio en el
primer render -- en el segundo 8,28 seguia el subtitulo que acababa en el
7,64. Por eso va fps antes que subtitles: asi cada fotograma de salida existe
de verdad y consulta el SRT por su cuenta.
"""
import argparse
import csv
import os
import subprocess
import sys

CSV = "hoja-montaje.csv"
SRT = "subtitulos.srt"
AUDIO = "david-elevenlabs/david-narracion.mp3"
EXTENSIONES = (".png", ".jpg", ".jpeg", ".webp")
W, H, FPS = 1920, 1080, 30


def buscar(carpeta, n):
    for ext in EXTENSIONES:
        p = os.path.join(carpeta, f"{n:03d}{ext}")
        if os.path.exists(p):
            return p
    return None


def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("carpeta")
    ap.add_argument("--zoom", action="store_true")
    ap.add_argument("--subtitulos", action="store_true")
    ap.add_argument("--salida", default="video-8-montado.mp4")
    ap.add_argument("--audio", default=AUDIO)
    args = ap.parse_args()

    filas = list(csv.DictReader(open(CSV, encoding="utf-8")))
    rutas, faltan = [], []
    for f in filas:
        n = int(f["plano"])
        p = buscar(args.carpeta, n)
        (rutas.append((p, float(f["dur"]))) if p else faltan.append(n))

    if faltan:
        print(f"::error:: faltan {len(faltan)} imágenes: "
              f"{', '.join(f'{n:03d}' for n in faltan[:15])}"
              f"{'...' if len(faltan) > 15 else ''}")
        sys.exit(1)
    if not os.path.exists(args.audio):
        print(f"::error:: no encuentro el audio: {args.audio}")
        sys.exit(1)

    # Encajar cualquier tamaño en 1920x1080 sin deformar: se escala hasta
    # cubrir y se recorta el sobrante. Si se usara pad quedarian franjas
    # negras cuando una imagen no salga exactamente en 16:9.
    encaje = (f"scale={W}:{H}:force_original_aspect_ratio=increase,"
              f"crop={W}:{H},setsar=1")

    # El filtro subtitles lee el SRT y aplica este estilo. Va SIEMPRE al final
    # de la cadena: si se pusiera antes del escalado, el texto se escalaria y
    # recortaria con la imagen.
    quemar = ""
    if args.subtitulos:
        if not os.path.exists(SRT):
            print(f"::error:: no encuentro los subtítulos: {SRT}")
            sys.exit(1)
        estilo = ("FontName=DejaVu Sans,Fontsize=22,Bold=1,"
                  "PrimaryColour=&H00FFFFFF,OutlineColour=&H00000000,"
                  "BorderStyle=1,Outline=2,Shadow=1,"
                  "Alignment=2,MarginV=60")
        quemar = f",subtitles={SRT}:force_style='{estilo}'"

    if args.zoom:
        # zoompan trabaja por fotogramas: cada plano dura dur*FPS frames y el
        # zoom va de 1.00 a 1.08 a lo largo de ese tramo, centrado.
        trozos = []
        for i, (p, d) in enumerate(rutas):
            fr = max(1, round(d * FPS))
            trozos.append(
                f"[{i}:v]{encaje},zoompan=z='1+0.08*on/{fr}':d={fr}:"
                f"x='iw/2-(iw/zoom/2)':y='ih/2-(ih/zoom/2)':s={W}x{H}:fps={FPS}[v{i}]"
            )
        cadena = ";".join(trozos)
        cadena += ";" + "".join(f"[v{i}]" for i in range(len(rutas)))
        cadena += f"concat=n={len(rutas)}:v=1:a=0"
        cadena += f"{quemar}[vid]" if quemar else "[vid]"

        # FALLO QUE HUBO AQUI: "-t 1" limitaba a un segundo fijo la entrada de
        # CADA imagen, sin importar cuanto durara su plano de verdad (2 a 6 s
        # segun la hoja). zoompan necesita fotogramas de entrada durante TODA
        # la duracion del plano; en cuanto se agotaban al segundo 1, ffmpeg se
        # quedaba repitiendo el ultimo fotograma disponible -- el plano se veia
        # congelado el resto de su duracion. Cada imagen debe entrar el tiempo
        # que le toca a SU plano, no un segundo parejo para todas.
        cmd = ["ffmpeg", "-y"]
        for p, d in rutas:
            cmd += ["-loop", "1", "-t", f"{d:.3f}", "-i", p]
        cmd += ["-i", args.audio, "-filter_complex", cadena, "-map", "[vid]",
                "-map", f"{len(rutas)}:a"]
    else:
        # Demuxer concat: mucho mas rapido y sin limite practico de entradas.
        lista = "lista-concat.txt"
        with open(lista, "w", encoding="utf-8") as fh:
            for p, d in rutas:
                fh.write(f"file '{os.path.abspath(p)}'\nduration {d:.3f}\n")
            # El demuxer ignora la duracion del ultimo fichero salvo que se
            # repita la linea file: sin esto el plano 95 dura un fotograma.
            fh.write(f"file '{os.path.abspath(rutas[-1][0])}'\n")
        cmd = ["ffmpeg", "-y", "-f", "concat", "-safe", "0", "-i", lista,
               "-i", args.audio,
               "-vf", f"{encaje},fps={FPS}{quemar}", "-r", str(FPS)]

    cmd += ["-c:v", "libx264", "-preset", "medium", "-crf", "18",
            "-pix_fmt", "yuv420p", "-c:a", "aac", "-b:a", "192k",
            "-shortest", args.salida]

    total = sum(d for _, d in rutas)
    print(f"{len(rutas)} imágenes, {total:.2f}s, "
          f"zoom={'sí' if args.zoom else 'no'}, "
          f"subtítulos={'quemados' if args.subtitulos else 'no'}")
    print(f"render -> {args.salida}\n")
    subprocess.run(cmd, check=True)
    print(f"\nlisto: {args.salida}")


if __name__ == "__main__":
    main()
